mono/metadata/gc.c

```diff
@@ -681,6 +682,31 @@ ves_icall_System_GCHandle_GetTarget (guint32 handle)
 	return mono_gchandle_get_target (handle);
 }
 
+static gboolean
+is_object_pinnable (MonoObject *obj)
```

Comment:
> Why is it checking for MONO_TYPE_CLASS?

This corresponds to the `CannotAllocAPinnedGCHandleToAClass` test. I don't believe that it is valid to allocate a pinned GC handle to a class in any case.

> Why it is giving special treatment to ARRAY and SZARRAY in the array case?

Arrays of arrays are not pinnable. This corresponds to the `CannotAllocAPinnedGCHandleToAnArrayOfIntArrays` test.

> The tests need to cover the positive tests as well.

I did not add positive test cases because the existing behavior is to always allowing pinning of GC handles. I only added tests for the changed behaviors. I can add positive tests, but do some already exist?